pf/internal/convert/schema_context.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package convert
+
+import (
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/walk"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+)
+
+type schemaMapContext struct {
```

Comment:
I can talk through this; looking over this again it looks like awkward glue code again; I wanted to take this glue code out of the main converter module code.

Looks like these types step down Element or GetAttr (which is the same thing as SchemaPath elements) to co-locate the schema information.  Instead of writing it like this, we could instead hold on to top-level shim.SchemaMap and simply walk down the SchemaPath  and always lookup from the top using walk.LookupSchemaMapPath.

It seems like in many places we want a record that holds shim.SchemaMap and a matching *tfbridge.SchemaInfo together; or else a Schema and matching tfbridge.SchemaInfo together. This is not the only place.. these pieces of information naturally join. I wonder if we could name it better and reuse.

Reply:
I agree that this should be a separate package. 

Maybe something like `bridged.SchemaMap`?

Reply:
Hm yes or SchemaWithInfo and SchemaMapWithInfo ? This PR is too big to take this on but perhaps another time we're touching something like this.